### ISSUE TITLE
Use Nullable Type Attributes for Properties/Fields where Initialization is Outside the Constructor

### ISSUE THREAD

**Issue by MarkMichaelis** [IntelliTect/CodingStandards#89]
Given properties that the compiler can't detect initialization for, such as MSTest's `TestContext` or a calculated `Name` property that splits the assignment to `FirstName` and `LastName`, use nullable attributes.
Examples:
```
[DisallowNull][NotNull]
public TestContext? TestContext {get;set;}
```

and
```
Person(string name) { Name = name; }
[DisallowNull][NotNull]
public string? FirstName {get;set;}
[DisallowNull][NotNull]
public string? FirstName {get;set;}
public string Name 
{
    get => $"{FirstName} {LastName}";
    set => { ... }
}
```

Proposal:
Create a code analysis rule that adds the attributes rather than creates generates a pragma.

**Comment by Keboo** [comment]
This is not currently possible given that the attributes are not available on all frameworks, so the analyzer would then possibly generate invalid C#. I propose we blog this and close this issue.

**Comment by MarkMichaelis** [comment]
Yes to blogging it.  I will do take that on.

If this is the right thing to do when available, then we just have to include a qualifier (3.0 or later) in the guideline, no?

**Comment by Keboo** [comment]
Yes, though we should follow up about getting those attributes distributed.

**Comment by MarkMichaelis** [comment]
Exactly!
